runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,250 @@
+		fomrobject_t *endPtr = limitPtr;
+		if (!GC_ObjectScanner::isIndexableObjectNoSplit(flags) && (splitAmount != 0)) {
+			Assert_MM_unreachable();
+			endPtr = scanPtr + (splitAmount * elementSize);
+			if (endPtr > limitPtr) {
+				endPtr = limitPtr;
+			}
+		}
+
+		fprintf(stderr, "!!! Flattened array: ");
+		struct J9UTF8* name = J9ROMCLASS_CLASSNAME(elementClass->romClass);
+		fprintf(stderr, "%.*s ", J9UTF8_LENGTH(name), J9UTF8_DATA(name));
+		fprintf(stderr, "elementCount=%ld, ", sizeInElements);
+		fprintf(stderr, "elementSize=%ld, ", elementSize);
+		fprintf(stderr, "(%p, %p)", scanPtr, endPtr);
+		fprintf(stderr, "\n");
+
+		new(objectScanner) GC_FlattenedArrayObjectScanner(env, objectPtr, basePtr, limitPtr, scanPtr, endPtr, 0, elementSize, elementClass, flags);
+		objectScanner->initialize(env, elementClass);
+		if (0 != startIndex) {
+			objectScanner->clearHeadObjectScanner();
+		}
+		return objectScanner;
+	}
+
+	MMINLINE uintptr_t getBytesRemaining() { return sizeof(fomrobject_t) * (_endPtr - _scanPtr); }
+
+	/**
+	 * @param env The scanning thread environment
+	 * @param allocSpace pointer to space within which the scanner should be instantiated (in-place)
+	 * @param splitAmount The maximum number of array elements to include
+	 * @return Pointer to split scanner in allocSpace
+	 */
+	GC_IndexableObjectScanner *
+	splitTo(MM_EnvironmentBase *env, void *allocSpace, uintptr_t splitAmount)
+	{
+		Assert_MM_unimplemented();
+	}
+
+	/**
+	 * Return base pointer and slot bit map for next block of contiguous slots to be scanned. The
+	 * base pointer must be fomrobject_t-aligned. Bits in the bit map are scanned in order of
+	 * increasing significance, and the least significant bit maps to the slot at the returned
+	 * base pointer.
+	 *
+	 * @param[out] scanMap the bit map for the slots contiguous with the returned base pointer
+	 * @param[out] hasNextSlotMap set this to true if this method should be called again, false if this map is known to be last
+	 * @return a pointer to the first slot mapped by the least significant bit of the map, or NULL if no more slots
+	 */
+	virtual fomrobject_t *
+	getNextSlotMap(uintptr_t *slotMap, bool *hasNextSlotMap)
+	{
+		fomrobject_t *result = GC_HeadlessMixedObjectScanner::getNextSlotMap(slotMap, hasNextSlotMap);
+		if (result != NULL) {
+			/* Ignore hasNextSlotMap from HeadLess, we want to always report that there is another element */
+			*hasNextSlotMap = true;
+			return result;
+		}
+
+		// No more slots in the current element, get the next element of the array
+		result = _indexableScanner.next();
+		if (result == NULL) {
+			// There are no elements in the array
+			*hasNextSlotMap = false;
+			return result;
+		}
+
+		// Reset the scanner to start scanning the next element
+		_mapPtr = result;
+		_endPtr = (fomrobject_t *)((uint8_t*)_mapPtr + _env->getExtensions()->mixedObjectModel.getSizeInBytesWithoutHeader(_elementClazzPtr));
```

Comment:
make sense to store element size (and I believe **rounded**) in the class instead of call helper every time

Reply:
there is a dilemma. Rounded up element size might be stored (essentially including padding to element) but in this case next bit in array should be 0 (and probably it is) but it means this solution has a side effects. Alternatively padding size itself can be stored in the class and be used for address adjustment but not to pass to element iterator

Reply:
I think we let IndexableScanner use rounded up element size, and HeadLess should use unrounded element size.  For this line in particular we should use unrounded element size to make sure we are not scanning junk padding data inside the array element.

Reply:
Yes, store element size and padding size separately is much cleaner solution. 
However (just academically) in case we have now (32-bit padding required in case object size is not aligned to 64-bit but next element must be) we can walk away and never scan garbage at padding slot. The trick is the bit related to padding is located at the same word as other bits so it happen to be initialized to 0 always. But it is better not to rely on this obviously